File: app/src/main/java/org/deiverbum/app/data/entity/PaterOpusEntity.java
package org.deiverbum.app.data.entity;

import static org.deiverbum.app.utils.Constants.PATER_OPUS;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2022.2
 *
 * CREATE TABLE `obra` (
 * 	`obraId` INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
 * 	`obra` TEXT NOT NULL,
 * 	`liturgyName` TEXT NOT NULL DEFAULT '',
 * 	`subTitulo` TEXT DEFAULT NULL,
 * 	`fecha` INTEGER DEFAULT NULL,
 * 	`volumen` INTEGER DEFAULT NULL,
 * 	`editorial` TEXT DEFAULT NULL,
 * 	`ciudad` TEXT DEFAULT NULL,
 * 	`year` INTEGER DEFAULT NULL,
 * 	`padreFK` INTEGER NOT NULL DEFAULT 0,
 * 	`tipoFK` INTEGER NOT NULL DEFAULT 0,
 * 	`coleccionFK` INTEGER NOT NULL DEFAULT 0,
 * 	FOREIGN KEY(`padreFK`) REFERENCES `padre`(`padreId`) ON DELETE CASCADE ON UPDATE CASCADE,
 * 	UNIQUE (`obra`, `padreFK`, `volumen`)
 *  );
 */

@Entity(tableName = PATER_OPUS,
        foreignKeys =
                {
                        @ForeignKey(
                                entity = PaterEntity.class,
                                parentColumns = "paterID",
                                childColumns = "paterFK",
                                onDelete = ForeignKey.CASCADE,
                                onUpdate = ForeignKey.CASCADE)},
         indices = {@Index(value = {"opus","paterFK","volume"},unique = true)}
)
public class PaterOpusEntity {
    @NonNull
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "opusID")
    public Integer obraId=0;

    @NonNull
    @ColumnInfo(name = "opus")
    public String obra="";

    @NonNull
    @ColumnInfo(name = "liturgyName", defaultValue = "")
    public String liturgyName="";

    //@NonNull
    @ColumnInfo(name = "subTitle", defaultValue = "NULL")
    public String subTitulo="";

    //@NonNull
    @ColumnInfo(name = "volume", defaultValue = "NULL")
    public Integer volumen=0;

    @ColumnInfo(name = "date", defaultValue = "NULL")
    public Integer fecha=0;

    //@NonNull
    @ColumnInfo(name = "editorial", defaultValue = "NULL")
    public String editorial="";

    //@NonNull
    @ColumnInfo(name = "city", defaultValue = "NULL")
    public String ciudad;

    //@NonNull
    @ColumnInfo(name = "year", defaultValue = "NULL")
    public Integer year;

    @NonNull
    @ColumnInfo(name = "paterFK", defaultValue = "0")
    public Integer padreFK=0;

    @NonNull
    @ColumnInfo(name = "typeFK", defaultValue = "0")
    public Integer tipoFK=0;

    @NonNull
    @ColumnInfo(name = "collectionFK", defaultValue = "0")
    public Integer coleccionFK=0;

    @NonNull
    public String getObra() {
        return obra;
    }
}

File: app/src/main/java/org/deiverbum/app/data/entity/PaterOpusEntity.kt
package org.deiverbum.app.data.entity

import androidx.room.*
import androidx.room.ForeignKey.Companion.CASCADE
import org.deiverbum.app.utils.Constants

/**
 * Entidad que representa a la tabla {@value org.deiverbum.app.utils.Constants#PATER_OPUS}.
 *
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
@Entity(
    tableName = Constants.PATER_OPUS,
    foreignKeys = [ForeignKey(
        entity = PaterEntity::class,
        parentColumns = arrayOf("paterID"),
        childColumns = arrayOf("paterFK"),
        onDelete = CASCADE,
        onUpdate = CASCADE
    )],
    indices = [Index(value = ["opusName", "paterFK", "volume"], unique = true)]
)
class PaterOpusEntity {
    @JvmField
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "opusID")
    var obraId = 0

    @JvmField
    @ColumnInfo(name = "opusName")
    var opusName = ""

    @JvmField
    @ColumnInfo(name = "liturgyName", defaultValue = "")
    var liturgyName = ""

    //@NonNull
    @JvmField
    @ColumnInfo(name = "subTitle", defaultValue = "NULL")
    var subTitle : String? = null

    //@NonNull
    @JvmField
    @ColumnInfo(name = "volume", defaultValue = "NULL")
    var volumen : Int? = null

    @JvmField
    @ColumnInfo(name = "opusDate", defaultValue = "NULL")
    var opusDate : Int? = null

    //@NonNull
    @JvmField
    @ColumnInfo(name = "editorial", defaultValue = "NULL")
    var editorial : String? = null

    //@NonNull
    @JvmField
    @ColumnInfo(name = "city", defaultValue = "NULL")
    var ciudad: String? = null

    //@NonNull
    @JvmField
    @ColumnInfo(name = "opusYear", defaultValue = "NULL")
    var opusYear: Int? = null

    @JvmField
    @ColumnInfo(name = "paterFK", defaultValue = "0", index = true)
    var padreFK = 0

    @JvmField
    @ColumnInfo(name = "typeFK", defaultValue = "0")
    var typeFK = 0

    @JvmField
    @ColumnInfo(name = "collectionFK", defaultValue = "0")
    var collectionFK = 0
}